core/services/gateway/network/wsconnection.go

```diff
@@ -0,0 +1,161 @@
+package network
+
+import (
+	"context"
+	"errors"
+	"sync/atomic"
+
+	"github.com/gorilla/websocket"
+)
+
+// WSConnectionWrapper is a websocket connection abstraction that supports re-connects.
+// I/O is separated from connection management:
+//   - component doing writes can use the thread-safe Write() method
+//   - component doing reads can listen on the ReadChannel()
+//   - component managing connections can listen to connection-closed channels and call Restart()
+//     to swap the underlying connection object
+//
+// The Wrapper can be used by a server expecting long-lived connections from a given client,
+// as well as a client maintaining such long-lived connection with a given server.
+// This fits the Gateway very well as servers accept connections only from a fixed set of nodes
+// and conversely, nodes only connect to a fixed set of servers (Gateways).
+//
+// The concept of "pumps" is borrowed from https://github.com/smartcontractkit/wsrpc
+// All methods are thread-safe.
+type WSConnectionWrapper interface {
+	// Update underlying connection object. Return a channel that gets an error on connection close.
+	// Cannot be called after Shutdown().
+	Restart(newConn *websocket.Conn) <-chan error
+
+	Write(ctx context.Context, msgType int, data []byte) error
+
+	ReadChannel() chan ReadItem
+
+	Shutdown()
+}
+
+type wsConnectionWrapper struct {
+	conn atomic.Pointer[websocket.Conn]
+
+	writeCh    chan writeItem
+	readCh     chan ReadItem
+	shutdownCh chan struct{}
+	shutdown   atomic.Bool
+}
+
+type ReadItem struct {
+	MsgType int
+	Data    []byte
+}
+
+type writeItem struct {
+	MsgType int
+	Data    []byte
+	ErrCh   chan error
+}
+
+var _ WSConnectionWrapper = (*wsConnectionWrapper)(nil)
+
+var (
+	ErrNoActiveConnection = errors.New("no active connection")
+	ErrWrapperShutdown    = errors.New("wrapper shutting down")
+)
+
+func NewWSConnectionWrapper() WSConnectionWrapper {
+	cw := &wsConnectionWrapper{
+		writeCh:    make(chan writeItem),
+		readCh:     make(chan ReadItem),
+		shutdownCh: make(chan struct{}),
+	}
+	// write pump runs until Shutdown() is called
+	go cw.writePump()
+	return cw
+}
+
+// Restart
+//  1. replaces the underlying connection and shuts the old one down
+//  2. starts a new read goroutine that pushes received messages to readCh
+//  3. returns channel that closes when connection closes
+func (c *wsConnectionWrapper) Restart(newConn *websocket.Conn) <-chan error {
+	oldConn := c.conn.Swap(newConn)
+
+	if oldConn != nil {
+		oldConn.Close()
+	}
+	if newConn == nil {
+		return nil
+	}
+	closeCh := make(chan error, 1)
+	// readPump goroutine is tied to the lifecycle of the underlying conn object
+	go c.readPump(newConn, closeCh)
+	return closeCh
+}
+
+func (c *wsConnectionWrapper) Write(ctx context.Context, msgType int, data []byte) error {
+	errCh := make(chan error, 1)
```

Comment:
`defer close(errCh)`

Reply:
And, run all tests with race detector enabled: https://go.dev/doc/articles/race_detector

Reply:
I can only close from writePump(). If I try closing in Write() then I'm risking writing to a closed channel.
Test passes with -race (I believe CI is running with that flag, no?)